base/multi.jl

```diff
@@ -1024,33 +1037,55 @@ function message_handler_loop(r_stream::IO, w_stream::IO)
     end
 end
 
-handle_msg(msg::CallMsg{:call}, r_stream, w_stream) = schedule_call(msg.response_oid, ()->msg.f(msg.args...; msg.kwargs...))
-function handle_msg(msg::CallMsg{:call_fetch}, r_stream, w_stream)
+function process_hdr(s, validate_cookie)
+    if validate_cookie
+        cookie = read(s, HDR_COOKIE_LEN)
+        self_cookie = cluster_cookie()
+        for i in 1:HDR_COOKIE_LEN
+            if UInt8(self_cookie[i]) != cookie[i]
+                error("Process($(myid())) - Invalid connection credentials sent by remote.")
+            end
+        end
+    end
+
+    # When we have incomaptible julia versions trying to connect to each other,
+    # and can be detected, raise an appropriate error.
+    # For now, just return the version.
+    return VersionNumber(strip(String(read(s, HDR_VERSION_LEN))))
```

Comment:
All versions with the same major version (for major ver 0, all versions with the same minor version) must be compatible. Being aware of the remote version helps us implement bug/security fixes that need changes in the serialization / messaging protocol.

For incompatible versions we will print a nice error message. Currently it is moot since versions 0.4 and below will probably just segfault out. Applicable for versions 0.5 and later.